### DISCUSSION TITLE
Application Insights does not work with native fetch

### DISCUSSION THREAD

**Discussion by cristobal** [nodejs/undici#2150] [General]
Application Insights has an auto collect feature that collects http(s) request calls done by an application. However this no longer works with the native fetch implementation.

The reason for this is that undici neither use `http` nor `https` layers, but rather is a re-implementation. This is the same reason that libraries such as `nock` can't intercept and mock request done via native fetch.

So are there any plans on exposing a surface that libraries such that [**nock**](https://github.com/nock/nock) and [**application insights**](https://github.com/microsoft/ApplicationInsights-node.js), can listen to and/or intercept?

## What are the alternatives?
1. A possibility is to rather use [**node-fetch**](https://github.com/node-fetch/node-fetch) instead (low effort)
2. Create a proxy function that wraps the native fetch implementation, and manually track calls done by the native-fetch implementation to Application Insights.

**Comment by mcollina** [comment]
This issue is better placed on the  Application Insight repository.

Undici supports diagnostics channel: https://undici.nodejs.org/#/docs/api/DiagnosticsChannel, which was designed to be enough for the Application Insight use case.

In case it's insufficient, we are happy to work with anybody to add the metadata that is missing.

Regarding nock, undici ships with its own mocking system.

**Reply by cristobal** [reply]
Thanks for the reply @mcollina , will rather create and issue at the application insights repository as you said.

In regards to nock i am aware of the mocking system, i have been following this issue on nock's side [nock/nock/issues/2183](https://github.com/nock/nock/issues/2183).

I hope you can soon land `undici` as a native package in node soon 😊 👋🏽 
That would be nice 👌🏽
